### PR TITLE
blockchain: clearly distinguish vm error from other errors

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
- klaytn code distinguishes the vm error and other errors, but the code was not clear enough. By bringing the ethereum PR, it clearly distinguish vm error from other errors.
- vm error is an internal evm errror and we record this kind of error at the transaction receipt. If not, we don't record it, instead the error is returned immediately(call) or causes a bad block error(sendTransaction).

- This PR is a refactoring, so the output of the rpc does not changes compared to the dev branch except revert reason.
  - dev: There's no 'data' field when there's revert reason.
  - PR: There's 'data' field when there's revert reason. It is a hexutil encoded value of the revert reason.
```
➜  klaytn_private curl -H "Content-Type: application/json" --data '{...}, "latest"], "id": 1}' http://localhost:8559 
{"jsonrpc":"2.0","id":1,"error":{"code":3,"message":"evm: execution reverted: new value should not be same as old value","data":"0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000296e65772076616c75652073686f756c64206e6f742062652073616d65206173206f6c642076616c75650000000000000000000000000000000000000000000000"}}
```

- This PR reorganizes the return values of `TransitionDb()` and `ApplyMessage()`

| Before PR                 | After PR                                                 | Meaning                                  |
|---------------------------|----------------------------------------------------------|------------------------------------------|
| ret []byte                | ExecutionResult.ReturnData []byte <br>~ result.Return(), result.Revert()      | Execution result (solidity return values)                    |
| gas uint64                | ExecutionResult.UsedGas uint64                           | Gas used                                 |
| kerror.Status uint        | ExecutionResult.VmExecutionStatus uint <br>~ result.Unwrap(), result.Failed() | ex: Success(1), OutOfGas(7), Reverted(9) |
| kerror.ErrTxInvalid error | err error                                                | ex: "insufficient balance of the sender" |

In turn, some internal functions' return values are reorganized as well.

| Before PR                                            | After PR                                   |
|------------------------------------------------------|--------------------------------------------|
| EthDoCall() -><br> returnData, gas, status, err          | EthDoCall() -><br> ExecutionResult, err        |
| DoCall() -><br> returnData, gas, compCost, status, error | DoCall() -><br> ExecutionResult, compCost, err |

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

### CODE REVIEW

File: blockchain/vm/evm.go [hunk 1]
```diff
@@ -304,8 +304,10 @@ func (evm *EVM) CallCode(caller types.ContractRef, addr common.Address, input []
 	if evm.depth > int(params.CallCreateDepth) {
 		return nil, gas, ErrDepth // TODO-Klaytn-Issue615
 	}
-	// Fail if we're trying to transfer more than the available balance
-	if !evm.CanTransfer(evm.StateDB, caller.Address(), value) {
+	// Note although it's noop to transfer X ether to caller itself. But
+	// if caller doesn't have enough balance, it would be an error to allow
+	// over-charging itself. So the check here is necessary.
+	if !evm.Context.CanTransfer(evm.StateDB, caller.Address(), value) {
```
Comment (kjeom) [inline blockchain/vm/evm.go:310]: How about making it evm.Context.CanTransfer or evm.CanTransfer to be consistent elsewhere(every callsite)?

File: accounts/abi/bind/backends/simulated_test.go [hunk 2]
```diff
@@ -397,25 +399,120 @@ func TestSimulatedBackend_TransactionByHash(t *testing.T) {
...
+			To:       &contractAddr,
+			Gas:      100000,
+			GasPrice: big.NewInt(0),
+			Value:    nil,
+			Data:     common.Hex2Bytes("e09fface"),
+		}, 21483, nil, nil},
+	}
+	for _, c := range cases {
+		got, err := sim.EstimateGas(context.Background(), c.message)
+		if c.expectError != nil {
```
Comment (kjeom) [inline accounts/abi/bind/backends/simulated_test.go:497]: confused a little bit. Is it matter if the expectError is nil or not?

if the expectedError is of the same type as the value returned by the function, I think the following code is enough
```go
if c.expectError != err {
    t.Fatalf("Expect error, want %v, got %v", c.expectError, err)
}
if c.expectData != nil {
    if err, ok := err.(*revertError); !ok {
        t.Fatalf("Expect revert error, got %T", err)
    } else if !reflect.DeepEqual(err.ErrorData(), c.expectData) {
        t.Fatalf("Error data mismatch, want %v, got %v", c.expectData, err.ErrorData())
    }
}
```

**Comment by kjeom** [comment]
I left a few minor reviews and this PR looks more clear than before. Thank you for hard work.

I think we can remove the duplicated code(`DoEstimateGas`) in the eth/klay namespace, 
but I think it would be nice to do this later while refactoring the API.

File: blockchain/state_transition.go [hunk 3]
```diff
@@ -243,50 +287,64 @@ func (st *StateTransition) preCheck() error {
...
+// - returndata:
+//      the returned data from evm
+// - vm execution status:
+//      indicates the execution result of a transaction. if the execution succeed, the status is 1.
+//      if it fails, the status indicates various **EVM** errors which abort the execution.
+//      e.g. ReceiptStatusErrOutOfGas, ReceiptStatusErrExecutionReverted
+//
+// However if any consensus issue encountered, return the error directly with
+// nil evm execution result.
+func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
```
[the remark this answers is not in the document]
Comment (yoomee1313) [reply blockchain/state_transition.go:303]: `st` stands for `StateTransition`, so I think `st` is better to stand the `StateTransition`.

File: blockchain/vm/evm.go (hunk 1 above)
Comment (yoomee1313) [reply blockchain/vm/evm.go:310]: https://github.com/klaytn/klaytn/pull/1763/commits/b9af23c7d5af383f5657c40a6f7957c74a405455. Changed the `create` to use `evm.Context.CanTransfer`.

File: blockchain/vm/internaltx_trace_json_test.go [hunk 4]
```diff
@@ -79,7 +80,7 @@ func TestInternalTxTrace_MarshalJSON(t *testing.T) {
 				Gas:     2971112,
 				GasUsed: 195,
 				Input:   "0x73b40a5c000000000000000000000000400de2e016bda6577407dfc379faba9899bc73ef0000000000000000000000002cc31912b2b0f3075a87b3640923d45a26cef3ee000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000064d79d8e6c7265636f76657279416464726573730000000000000000000000000000000000000000000000000000000000383e3ec32dc0f66d8fe60dbdc2f6815bdf73a988383e3ec32dc0f66d8fe60dbdc2f6815bdf73a98800000000000000000000000000000000000000000000000000000000000000000000000000000000",
-				Error:   errExecutionReverted,
+				Error:   errors.New("execution reverted"),
```
[the remark this answers is not in the document]
Comment (yoomee1313) [reply blockchain/vm/internaltx_trace_json_test.go]: It was mistake. Recovered to original code.

**Comment by yoomee1313** [comment]
@kjeom I'm sorry for the late. Thank you for your patience. I agree with you about the refactoring of `DoEstimate`. I'll reflect it in another PR cause this PR is already too big.

File: accounts/abi/bind/backends/simulated_test.go (hunk 2 above)
Comment (yoomee1313) [reply accounts/abi/bind/backends/simulated_test.go:497]: `c.expectError != err` doesn't work. It has to check `err.Error()`. But, if before calling Error(), it needs to check err is nil or not.

Comment (kjeom) [reply accounts/abi/bind/backends/simulated_test.go:497]: I got it, thanks

**Review by kjeom — APPROVED** [review]

**Review by blukat29 — APPROVED** [review]

**Review by jiseongnoh — APPROVED** [review]

**Pull request merged**